### ISSUE TITLE
`lfs migrate info` panic

### ISSUE THREAD

**Issue by recih** [git-lfs/git-lfs#5332]
**Describe the bug**
```bash
# git lfs migrate info --everything --fixup
migrate: Sorting commits: ..., done.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x28 pc=0x395287]

goroutine 1 [running]:
github.com/git-lfs/wildmatch/v2.(*Wildmatch).consume(0x0, {0xc0058b3200?, 0xc0002052c0?}, {0x98?})
        github.com/git-lfs/wildmatch/v2@v2.0.1/wildmatch.go:267 +0x27
github.com/git-lfs/wildmatch/v2.(*Wildmatch).Match(...)
        github.com/git-lfs/wildmatch/v2@v2.0.1/wildmatch.go:248
github.com/git-lfs/git-lfs/v3/git/gitattr.(*Tree).Applied(0xc000205180, {0xc0058b3200, 0x22})
        github.com/git-lfs/git-lfs/v3/git/gitattr/tree.go:95 +0xc5
github.com/git-lfs/git-lfs/v3/commands.migrateInfoCommand.func1({0xc0058b3200, 0x22}, 0xc0002052c0)
        github.com/git-lfs/git-lfs/v3/commands/command_migrate_info.go:131 +0xa5
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).rewriteBlob(0xc00013e4e0, {0xc000312720, 0x14, 0x27f01e?}, {0xc000605400, 0x14, 0x20}, {0xc0058b3200, 0x22}, 0xc0001b61e0, ...)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:469 +0xae
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).rewriteTree(0xc00013e4e0, {0xc000312720, 0x14, 0x30}, {0xc000605100, 0x14, 0x20}, {0xc00049f2d8, 0x13}, 0xc0001b61e0, ...)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:397 +0x869
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).rewriteTree(0xc00013e4e0, {0xc000312720, 0x14, 0x30}, {0xc000604fe0, 0x14, 0x20}, {0xc0004232c0, 0xc}, 0xc0001b61e0, ...)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:399 +0x805
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).rewriteTree(0xc00013e4e0, {0xc000312720, 0x14, 0x30}, {0xc002510720, 0x14, 0x20}, {0xc000423cf6, 0x5}, 0xc0001b61e0, ...)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:399 +0x805
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).rewriteTree(0xc00013e4e0, {0xc000312720, 0x14, 0x30}, {0xc000ed14d0, 0x14, 0x30}, {0x0, 0x0}, 0xc0001b61e0, ...)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:399 +0x805
github.com/git-lfs/git-lfs/v3/git/githistory.(*Rewriter).Rewrite(0xc00013e4e0, 0xc00007c3c0)
        github.com/git-lfs/git-lfs/v3/git/githistory/rewriter.go:247 +0x91b
github.com/git-lfs/git-lfs/v3/commands.migrate({0xc0002840e0, 0x0, 0x2}, 0x1?, 0xc0002a6000?, 0x70?)
        github.com/git-lfs/git-lfs/v3/commands/command_migrate.go:78 +0x8f
github.com/git-lfs/git-lfs/v3/commands.migrateInfoCommand(0xc000122000?, {0xc0002840e0, 0x0, 0x2})
        github.com/git-lfs/git-lfs/v3/commands/command_migrate_info.go:118 +0x68b
github.com/spf13/cobra.(*Command).execute(0xc000122000, {0xc0002840a0, 0x2, 0x2})
        github.com/spf13/cobra@v0.0.3/command.go:766 +0x663
github.com/spf13/cobra.(*Command).ExecuteC(0xc000122780)
        github.com/spf13/cobra@v0.0.3/command.go:852 +0x2e5
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v0.0.3/command.go:800
github.com/git-lfs/git-lfs/v3/commands.Run()
        github.com/git-lfs/git-lfs/v3/commands/run.go:108 +0x307
main.main()
        github.com/git-lfs/git-lfs/v3/git-lfs.go:34 +0xe5
```

**To Reproduce**
just run `git lfs migrate info --everything --fixup`

**Expected behavior**
git migrate info print successfully.

**System environment**
tested on multiple platforms, including Win10, WSL(Ubuntu 20.04), Debian 10

**Output of `git lfs env`**
```
git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3; git 77deabdf)
git version 2.40.0.windows.1

Endpoint=****
  SSH=****
LocalWorkingDir=D:\projects\demo_project
LocalGitDir=D:\projects\demo_project\.git
LocalGitStorageDir=D:\projects\demo_project\.git
LocalMediaDir=D:\projects\demo_project\.git\lfs\objects
LocalReferenceDirs=
TempDir=D:\projects\demo_project\.git\lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=D:\projects\demo_project\.git\lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**
It appears there may be some issues within my Git repository. However, it is a private repository,  and can not be public. 
So I need some help to providing more info.

**Comment by chrisd8088** [comment]
Very interesting -- thank you for the report and the stacktrace!  I see what the problem is here; it comes from the fact that when we parse a `.gitattributes` file which contains a macro definition, we [create](https://github.com/git-lfs/git-lfs/blob/bea0287cdd3acbc0aa9cdf67ae09b6843d3ffcf0/git/gitattr/attr.go#L143) a record of the line containing the macro in which the `Pattern` element is `nil`, because we only [set](https://github.com/git-lfs/git-lfs/blob/bea0287cdd3acbc0aa9cdf67ae09b6843d3ffcf0/git/gitattr/attr.go#L134-L139) an internal variable if a pattern is defined in the line.  We then expect to be able to [call](https://github.com/git-lfs/git-lfs/blob/bea0287cdd3acbc0aa9cdf67ae09b6843d3ffcf0/git/gitattr/tree.go#L95) the `Pattern` element's `Match()` method, and that crashes when the first access of the receiver argument is made.

Here's a reproduction case, for reference:
```
$ git init test
$ cd test
$ echo "[attr]foo foo" >.gitattributes
$ echo foo >foo.bin
$ git add .gitattributes foo.bin
$ git commit -m init
$ git lfs migrate info --fixup
panic: runtime error: invalid memory address or nil pointer dereference         
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x11843c7]

goroutine 1 [running]:
github.com/git-lfs/wildmatch/v2.(*Wildmatch).consume(0x0, {0xc00011c550?, 0xc00011a860?}, {0x80?})
	github.com/git-lfs/wildmatch/v2@v2.0.1/wildmatch.go:267 +0x27
github.com/git-lfs/wildmatch/v2.(*Wildmatch).Match(...)
	github.com/git-lfs/wildmatch/v2@v2.0.1/wildmatch.go:248
github.com/git-lfs/git-lfs/v3/git/gitattr.(*Tree).Applied(0xc00011a800, {0xc00011c550, 0x7})
	github.com/git-lfs/git-lfs/v3/git/gitattr/tree.go:95 +0xc5
...
```

This should be relatively straightforward to resolve, knowing the problem, but I'll think about whether we want to always define the `Pattern` element or be more cautious about invoking it.  Regardless, thanks again for the report!

**Comment by recih** [comment]
I can provide my `.gitattributes` file here, which is inspired from https://hextantstudios.com/unity-gitattributes/
I hope it's useful for fixing this problem.

Also, I found that `git lfs migrate import --everything --fixup` gets stuck on some files. Could it be related to the same problem?
```
# .gitattributes for Unity Projects
# https://hextantstudios.com/unity-gitattributes/

# Macro for Unity YAML-based asset files.
[attr]unityyaml -text merge=unityyamlmerge diff

# Macro for all binary files that should use Git LFS.
[attr]lfs -text filter=lfs diff=lfs merge=lfs

# Default to auto-normalized line endings.
* text=auto

# Code
*.cs text diff=csharp

# Unity Text Assets
*.meta unityyaml
*.unity unityyaml
*.asset unityyaml
*.prefab unityyaml
*.mat unityyaml
*.anim unityyaml
*.controller unityyaml
*.overrideController unityyaml
*.physicMaterial unityyaml
*.physicsMaterial2D unityyaml
*.playable unityyaml
*.mask unityyaml
*.brush unityyaml
*.flare unityyaml
*.fontsettings unityyaml
*.guiskin unityyaml
*.giparams unityyaml
*.renderTexture unityyaml
*.spriteatlas unityyaml
*.terrainlayer unityyaml
*.mixer unityyaml
*.shadervariants unityyaml
*.preset unityyaml
*.asmdef -text diff

# Unity Binary Assets
*.cubemap lfs
*.unitypackage lfs
# Note: Unity terrain assets must have "-Terrain" suffix.
*-[Tt]errain.asset -unityyaml lfs
# Note: Unity navmesh assets must have "-NavMesh" suffix.
*-[Nn]av[Mm]esh.asset -unityyaml lfs
[Ll]ighting[Dd]ata.asset -unityyaml binary

# Image
*.jpg lfs
*.jpeg lfs
*.png lfs
*.apng lfs
*.atsc lfs
*.gif lfs
*.bmp lfs
*.exr lfs
*.tga lfs
*.tiff lfs
*.tif lfs
*.iff lfs
*.pict lfs
*.dds lfs
*.xcf lfs
*.leo lfs
*.kra lfs
*.kpp lfs
*.clip lfs
*.webm lfs
*.webp lfs
*.svg lfs
*.svgz lfs
*.psd lfs
*.afphoto lfs
*.afdesign lfs

# Audio
*.mp3 lfs
*.ogg lfs
*.wav lfs
*.aiff lfs
*.aif lfs
*.mod lfs
*.it lfs
*.s3m lfs
*.xm lfs

# Video
*.mov lfs
*.avi lfs
*.asf lfs
*.mpg lfs
*.mpeg lfs
*.mp4 lfs
*.flv lfs
*.ogv lfs
*.wmv lfs

# 3D
*.fbx lfs
*.obj lfs
*.max lfs
*.blend lfs
*.blender lfs
*.dae lfs
*.mb lfs
*.ma lfs
*.3ds lfs
*.dfx lfs
*.c4d lfs
*.lwo lfs
*.lwo2 lfs
*.abc lfs
*.3dm lfs
*.bin lfs
*.glb lfs

# Executables
*.exe lfs
*.dll lfs
*.so lfs
*.pdb lfs
*.mdb lfs

# Packaging
*.zip lfs
*.7z lfs
*.gz lfs
*.rar lfs
*.tar lfs

# =========================================================================
# Project custom attribute settings
*.bundle lfs
*.model lfs
```

**Comment by chrisd8088** [comment]
Thanks!  I see a couple of macro definitions in your `.gitattributes` and so I'm pretty certain those are what's triggering this bug, just as in my reproduction case.

> Also, I found that `git lfs migrate import --everything --fixup` gets stuck on some files

Yes, I noticed that too even with my extremely simple reproduction case, so I'll look into that at the same time.

**Comment by chrisd8088** [comment]
I've had a bit of time to look further at this problem, and can report that the reason `git lfs migrate import --fixup` hangs where `git lfs migrate info --fixup` crashes is that there's a subtle difference in how they handle logging, introduced in commit 82169f318dcc1824c4c19d220081f5ddb35dd3ea of PR #2610, so that the former command [closes](https://github.com/git-lfs/git-lfs/blob/1efc16b424ea1f88bfbcfe81a667bcfb2cdba3d5/commands/command_migrate_import.go#L32) the log in a `defer` statement while the latter command [closes](https://github.com/git-lfs/git-lfs/blob/1efc16b424ea1f88bfbcfe81a667bcfb2cdba3d5/commands/command_migrate_info.go#L192) it directly.  The consequence is that although both commands' primary goroutines crash in the same way at the same spot, in the `git lfs migrate import` case the deferred log closing [introduces](https://github.com/git-lfs/git-lfs/blob/1efc16b424ea1f88bfbcfe81a667bcfb2cdba3d5/tasklog/log.go#L115) a wait on a `WaitGroup`, which never completes as there are no more log messages being generated and so the counter is never decremented to zero.  At least, that's my initial understanding of the situation.

So I'm also doing a review of how the logging functions are used and how to avoid hanging on error, which is what happens now.
